### PR TITLE
Error resilience

### PR DESCRIPTION
**Pull request by gares**
This feature branch enables Coq to be resilient to errors while processing a file.
The error may either be a top level sentence or a tactic occurring in a "proof block".
The branch defines proof block detection and error-recovery code for bullets, { blocks } and the par: selector/terminator.

A simple file to try out the whole thing is provided in test-suite/interactive: one can skip over a broken command/tactic as long as he lands on a valid one.  Proof workers can recover from a broken sub-proof.  Commands that recover from an error (by giving up the sub goals) are marked as unsafe.
coqide

I'm mostly interested in having testing feedback on the feature, not much about about the code itself, that is rather "STM specific".
I'm also interested in some feedback on the API (in stm.mli) to declare "proof block delimiters", that should be understandable by ML plugin writers.  All proof block code uses that API.

Gory implementation details are explained in the individual commit messages.

@ale-f this branch may interest you!

### CODE REVIEW

File: stm/stm.ml [hunk 1]
```diff
@@ -512,38 +587,73 @@ end = struct (* {{{ *)
...
-        { (get_info id) with state = None; vcs_backup = None,None } in
+        { (get_info id) with state = Empty; vcs_backup = None,None } in
     let copy_info_w_state v id =
       Vcs_.set_info v id { (get_info id) with vcs_backup = None,None } in
+    let copy_error_bound_boxes v =
+      let nodes = Vcs_.Dag.all_nodes (Vcs_.dag v) in
+      let props =
+        Stateid.Set.fold (fun n pl -> Vcs_.property_of !vcs n @ pl) nodes [] in
+      let props = CList.sort_uniquize Vcs_.Dag.Property.compare props in
+      (* XXX check if par: works here, since we copy a single node *)
```
Comment (gares) [inline stm/stm.ml]: Comment is outdated, par: seems to work now

File: stm/stm.ml [hunk 2]
```diff
@@ -1794,6 +1981,61 @@ let log_processing_sync id name reason = log_string Printf.(sprintf
...
 
 let known_state ?(redefine_qed=false) ~cache id =
+ 
+  let error_absorbing_tactic id exn =
+    let boxes = VCS.box_of id in
+    let valid_boxes = CList.map_filter (function
+      | ErrorBound ({ stop } as decl, name) when Stateid.equal stop id ->
+          Some (decl, name)
+      | _ -> None) boxes in
+    assert(List.length valid_boxes < 2);
```
Comment (gares) [inline stm/stm.ml]: Put anomaly here. (each tactic can be the error absorber for 1 and only 1 box).
Also, the static detection of proof blocks could catch case this early an report the error precisely.

**Comment by gares** [comment]
Reminder fir myself: check if the code sending back system states from the worker to the master sends states using the compact representation when an error is absorbed by a proof block delimiter.  If it is not the case, refine the heuristic that chooses the representation.

**Comment by gares** [comment]
Just checked: states are correctly sent back

**Comment by JasonGross** [comment]
Nice!  So this means that I can work on later subgoals without using `shelve` or `Focus` if I'm using CoqIDE.

Could this also be used to give coqc a mode/flag where it gives as many error messages as it can, rather than making me fix one error before getting the next message?

**Comment by gares** [comment]
To your first question: yes, but only if your proofs are structured.  As you can see I now consider {} and bullets as error confinement boxes (when they focus a goal containing no evars.. ie if admitting the goal has no visible impact on other goals).

To question 2: yes, in principle it is, but this patch does not do it.  Actually I think that you can try it out by thinkering with the if-then-else I added in the the commit that says "... made optional".  In coqc async_proof_mode is ApOff also in master.  Bottom line I've not tried it, so I've no idea what happens at .vo generation time.  I guess it fails (badly).

**Comment by aspiwack** [comment]
Interesting. You seem not to have changed the kernel, so there is no actual improvement in the parallelism of a Coq document. Am I correctly guessing that when an error is fixed, proofs which follow in the document need to be rechecked?

**Comment by gares** [comment]
@aspiwack you are right wrt parallelism. We use processes, so dispatching sub-proofs (as we do with par: ) may be profitable only for very heavy sub proofs. So I'm not too sure it is worth it.  In this patch error confinement boxes only serve to absorb errors and continue, but yes, they could also be used to parallelize more (semantically, the identify independent sub proofs). 

wrt proofs being rechecked you are not right.  As in 8.5 a broken (future) proof term is "an exception"; by repairing it you replace the exception by a value (an actual proof term), that is checked by the kernel.  So no, you don't re-check what follows unless you modify something that is already a value (a finished proof).  The trick is that the only non-local effect of checking a proof are universe constraint (they sort of leak the "opacity").  The consistency of the graph is checked only when constraints are ready.  So you may incur in late error detection, but there is nothing to redo when you fix a proof.  In other words adding constraints is a commutative/associative operation, hence checking finished proof term is.
This patch makes a proof "breakable" in different/multiple ways, but does not change the general schema.  I'd love to have a real notion of dependencies among proofs/definitions etc, but I don't have it now, so any value being modified implies re-checking everything that follows it.

Hope it helps.

**Comment by aspiwack** [comment]
Oh my, I knew exactly what I meant, but it's certainly not quite what I said :-p .

> you are right wrt parallelism

I meant parallelism in the mathematical sense (as in, fewer « before » relationships between the lines). From the rest of your post, I understand the answer is that the confinement boxes add some parallelism within proofs. I don't think we should think too much about having more work parallelised (in the computing sense).

> wrt proofs being rechecked you are not right. As in 8.5 a broken (future) proof term is "an exception"; by repairing it you replace the exception by a value (an actual proof term), that is checked by the kernel. So no, you don't re-check what follows unless you modify something that is already a value (a finished proof).

I was thinking of your `banana` example. If you fix it (_e.g_ by replacing `true` with `1`), you may affect every proof below since `banana` may be used by an automatic tactic (and since its type is unknown and it is transparent). Is that accounted for?

**Comment by gares** [comment]
OK, by proof I mean "opaque proof" as in "Proof...Qed.  Not as in "Definition..Defined".  So yes, fixing banana invalidates all the following text.

**Comment by aspiwack** [comment]
Makes sense to me.

I do also think that this is a desirable feature.

**Pull request merged**